Preparation for transaction sequence testing. (Part 3)

jonathanknowles opened the pull request:
## Issue Number

[ADP-1538](https://input-output.atlassian.net/browse/ADP-1538)

## Summary

This PR adds the following transformation functions:
```hs
Tx  .mapAssetIds :: (AssetId   -> AssetId  ) -> Tx   -> Tx
UTxO.mapAssetIds :: (AssetId   -> AssetId  ) -> UTxO -> UTxO
Tx  .mapTxIds    :: (Hash "Tx" -> Hash "Tx") -> Tx   -> Tx
UTxO.mapTxIds    :: (Hash "Tx" -> Hash "Tx") -> UTxO -> UTxO

Tx  .removeAssetId :: Tx   -> AssetId -> Tx
UTxO.removeAssetId :: UTxO -> AssetId -> UTxO
```

## Motivation

Suppose we have a sequence of transactions that apply to an initial UTxO set:

```
utxo_0 -> tx_0_1 -> utxo_1
       -> tx_1_2 -> utxo_2
       -> tx_2_3 -> utxo_3
       
       -> ...    -> ...
       
       -> tx_i_j -> utxo_j
```

When generating such a transaction sequence, we generate values of `TxId` and `AssetId` at random, both with fairly large ranges of values, in order to achieve a certain balance between the _uniqueness_ of generated values and the _collision rate_ with existing values. This space scales with the QC size parameter.

However, suppose we discover a transaction sequence that is a **_counterexample_** to some property. We'd ideally like to shrink that counterexample as much as possible. Counterexamples with very long `TxId` and `AssetId` values can be very hard to read, especially if they have unprintable characters requiring escape sequences.

When shrinking, we'd like the ability to simplify the sets of `TxId`s and `AssetId`s that are in use, to be concise values like this:

```hs
[AssetId "A", AssetId "B", AssetId "C", ...]
[TxId    "1", TxId    "2", TxId    "3", ...]
```

In order to ensure referential integrity within a given transaction sequence, we need to apply any transformations to `TxId` and `AssetId` values consistently across the entire sequence.

The functions provided in this PR provide building blocks for this ability.

jonathanknowles committed:
Add function `TokenBundle.setQuantity`.

jonathanknowles committed:
Add function `Tx.txOutRemoveAssetId`.

jonathanknowles committed:
Add function `UTxO.removeAssetId`.

jonathanknowles committed:
Add function `Tx.txRemoveAssetId`.

jonathanknowles committed:
Add function `Tx.txOutAssetIds`.

jonathanknowles committed:
Add function `Tx.txAssetIds`.

jonathanknowles committed:
Add function `UTxO.assetIds`.

jonathanknowles committed:
Add function `UTxO.mapTxIds`.

jonathanknowles committed:
Add function `TokenMap.mapAssetIds`.

jonathanknowles committed:
Add function `TokenBundle.mapAssetIds`.

jonathanknowles committed:
Add function `Tx.txOutMapAssetIds`.

jonathanknowles committed:
Add function `Tx.txMapAssetIds`.

jonathanknowles committed:
Add function `UTxO.mapAssetIds`.

jonathanknowles committed:
Add function `UTxO.txIds`.

jonathanknowles committed:
Add function `Tx.txMapTxIds`.

jonathanknowles committed:
Add dependency on library `quickcheck-instances`.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -330,6 +334,14 @@ getQuantity = TokenMap.getQuantity . tokens
+-- If the given quantity is zero, the resultant bundle will not have an entry
+-- for the given asset.
+--
+setQuantity :: TokenBundle -> AssetId -> TokenQuantity -> TokenBundle
```
```suggestion
setQuantity :: AssetId -> TokenQuantity -> TokenBundle -> TokenBundle
```

sevanspowell replied:
Suggestion: for easier currying and to make it's nature as a transformation on `TokenBundle` more obvious.

sevanspowell replied:
i.e. `setQuantity` is also a "transformation" like `mapAssetIds`.

jonathanknowles replied:
I can definitely sympathise with the currying argument.

For `TokenBundle.setQuantity`, I've chosen to mirror the signature of `TokenMap.setQuantity`, which uses the same argument order.

We could change them both so that they follow the order in your suggestion, but we'd have to change quite a lot of client code.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -330,6 +334,14 @@ getQuantity = TokenMap.getQuantity . tokens
 
+-- | Sets the quantity associated with a given asset.
+--
+-- If the given quantity is zero, the resultant bundle will not have an entry
```
An interface where the difference between zero quantity and "existence" is directly observable like this smells a bit to me, but this isn't really that important.

jonathanknowles replied:
> An interface where the difference between zero quantity and "existence" is directly observable like this smells a bit to me, but this isn't really that important.

Definitely agree!

I would argue that this sentence (and others like it) should be removed, as:
- `TokenMap` and `TokenBundle` both aim to provide a **_total_** mapping from assets to values: it's _always_ possible to retrieve a value for a given asset.
- The fact that empty values are never stored in the internal map is an _implementation detail_. (Though it's a pretty important one.)

However, for this PR I thought it best to retain consistency with the original interface.

I hope eventually to revise the API in a separate PR. (Though this may become unnecessary, once we remove `TokenMap` and `TokenBundle` and replace them with Cardano API / ledger types.)

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -968,6 +980,54 @@ txSizeDistance (TxSize a) (TxSize b)
+    & over #resolvedCollateralInputs
+        (fmap (first (over #inputId f)))
+
+txRemoveAssetId :: Tx -> AssetId -> Tx
```
Just a suggestion:
```suggestion
txRemoveAssetId :: AssetId -> Tx -> Tx
```

jonathanknowles committed:
Fix composition in `UTxO.mapTxIds`

The `mapTxIds` function did not satisfy the composition property:

```hs
mapTxIds f (mapTxIds g u) == mapTxIds (f . g) u
```

If there's a key collision within the `TxIn` key set, then one `TxOut`
value must be dropped (since there's no sensible way to combine `TxOut`
values).

Therefore, we need a consistent way to choose between pairs of `TxOut`
values.

Since `TxOut` has an `Ord` instance, this commit uses the `min` function
to choose between `TxOut` values.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3313/files#r890806221

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:
```diff
@@ -257,6 +277,29 @@ receiveD :: UTxO -> UTxO -> (DeltaUTxO, UTxO)
+mapAssetIds f (UTxO u) = UTxO $ Map.map (txOutMapAssetIds f) u
+
+mapTxIds :: (Hash "Tx" -> Hash "Tx") -> UTxO -> UTxO
+mapTxIds f (UTxO u) = UTxO $ Map.mapKeys (over #inputId f) u
```
This doesn't satisfy the composition property:

```hs
mapTxIds f (mapTxIds g u) == mapTxIds (f . g) u
```

If there's a collision between two `TxIn` values, then one `TxOut` value must be dropped (since there's no way to combine `TxOut` values). Therefore, we need a consistent way to choose between pairs of `TxOut` values.

Since `TxOut` has an `Ord` instance, we could just choose the "smaller" value (using the `min` function).

jonathanknowles replied:
Fixed in 5b8935a0badd0b6508b5e2655aa27dc1eb0d844a.
